docs/source/diagrams/pbft-view-change.puml

```diff
@@ -0,0 +1,129 @@
+                        - <b>stable_checkpoint</b> <i>(seq_no of the last stabilized checkpoint),
+                        - <b>prepared</b> <i>(list of tuples BatchId(view_no, pp_view_no, pp_seq_no, pp_digest))
+                        - <b>preprepared</b> <i>(list of BatchId(view_no, pp_view_no, pp_seq_no, pp_digest))
+                        - <b>checkpoints</b> <i>(list of checkpoint messages)
+    end note
+
+    Node3 -> Node4: VIEW_CHANGE_ACK(VIEW_CHANGE(Node1))
+    deactivate Node3
+    Node2 -> Node4: VIEW_CHANGE_ACK(VIEW_CHANGE(Node1))
+    deactivate Node2
+    Node4 -> Node4: VIEW_CHANGE_ACK(VIEW_CHANGE(Node1))
+    deactivate Node4
+    note right Node4
+          <b>VIEW_CHANGE_ACK contains:
+                        - <b>view_no</b> <i>(future view number),
+                        - <b>name</b> <i>(name of the node that sent VIEW_CHANGE message)
+                        - <b>digest</b> <i>(digest of VIEW_CHANGE message)
+    end note
+end loop
+
+Node4 -> Node3: NEW_VIEW
+Node4 -> Node2: NEW_VIEW
+Node4 -> Node1: NEW_VIEW
+note right Node4
+    <b>NEW_VIEW contains:
+                        - <b>view_no</b> <i>(number of future view change),
+                        - <b>view_changes</b> <i>(list of tuples(sender name, digest of VIEW_CHANGE messages)
+                        - <b>checkpoint</b> <i>(checkpoint to be selected as stable)
+                        - <b>batches</b> <i>(list of BatchId(view_no, pp_view_no, pp_seq_no, pp_digest))
+end note
+
+
+opt If Node1 lost a VIEW_CHANGE message
+Node1 -> Node2: MESSAGE_REQUEST(VIEW_CHANGE)
+Node1 -> Node3: MESSAGE_REQUEST(VIEW_CHANGE)
+Node1 -> Node4: MESSAGE_REQUEST(VIEW_CHANGE)
+note right Node4
+    Node1 send requests for a VIEW_CHANGE message to all nodes
+    because in theory it could have a problem with a network connection
+    with the owner of the lost message
+end note
+
+Node2 -> Node1: MESSAGE_RESPONSE(VIEW_CHANGE)
+Node3 -> Node1: MESSAGE_RESPONSE(VIEW_CHANGE)
+Node4 -> Node1: MESSAGE_RESPONSE(VIEW_CHANGE)
+note right Node4
+    If Node1 receive a VIEW_CHANGE message from its owner
+    or from quorum (f+1) of nodes then Node1 used it
+end note
+end opt
+
+opt If Node1 lost a NEW_VIEW message or receive it from a non-primary node
+Node1 -> Node2: MESSAGE_REQUEST(NEW_VIEW)
+Node1 -> Node3: MESSAGE_REQUEST(NEW_VIEW)
+Node1 -> Node4: MESSAGE_REQUEST(NEW_VIEW)
+note right Node4
+    Node1 send requests for a NEW_VIEW message to all nodes
+    because in theory it could choose a wrong primary
+end note
+
+Node2 -> Node1: MESSAGE_RESPONSE(NEW_VIEW)
+Node3 -> Node1: MESSAGE_RESPONSE(NEW_VIEW)
+Node4 -> Node1: MESSAGE_RESPONSE(NEW_VIEW)
+note right Node4
+    If Node1 doesn't receive NewView by the primary and uses a quorum(f+1) of responses
+    it finish the view change and start a <b> catchup</b>
+end note
+end opt
+
+note over Node1, Node4
+  <b>Finish the View change</b>
+ - Stabilize the checkpoint
+ - Unstash 3pc messages waiting for a new view
```

Comment:
I think the sequence of steps needs to be the following:
- Stabilize the checkpoint from the NewView
     - If the Node doesn't have the checkpoint in the NewView - do nothing (TBD: start a catchup here)
- Order or re-order all 3PC batches corresponding to the NewView:
    - find a corresponding PrePrepare if it's received by a Node- 
    - request PrePrepare from other nodes if it's not fiound
    - apply all PrePrepares if they are not ordered yet 
    - send all Prepares and Commits even for already ordered PrePrepares
- Unstash 3pc messages waiting for a new view
- reset the monitor
- Master Primary can start sending new PrePrepares once all PrePrepares from the old view (from NewView msg) are applied
- Backup Primaries can start sending PrePrepares once Master Instance ordered all PrePrepares from the old view (from NewView msg)